resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -467,20 +526,97 @@ private boolean handleRedirect(
                         continue;
                     }
 
-                    nameServers.add(new InetSocketAddress(resolved, parent.dnsRedirectPort(resolved)));
-                    addNameServerToCache(authoritativeNameServer, resolved, r.timeToLive());
+                    final InetSocketAddress socketAddress = parent.newRedirectServerAddress(resolved);
+                    nameServers.add(socketAddress);
+
+                    addNameServerToCache(authoritativeNameServer, socketAddress, r.timeToLive());
+                }
+
+                // Process all unresolved nameservers as well.
+                for (;;) {
+                    AuthoritativeNameServer authoritativeNameServer = serverNames.handleNext();
+
+                    if (authoritativeNameServer == null) {
+                        break;
+                    }
+                    // These will be resolved on the fly if needed.
+                    final InetSocketAddress unresolved = InetSocketAddress.createUnresolved(
+                            authoritativeNameServer.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                    nameServers.add(unresolved);
+
+                    addNameServerToCache(authoritativeNameServer, unresolved, authoritativeNameServer.timeToLive());
                 }
 
                 if (!nameServers.isEmpty()) {
-                    query(parent.uncachedRedirectDnsServerStream(nameServers), 0, question,
-                          queryLifecycleObserver.queryRedirected(unmodifiableList(nameServers)), promise, null);
-                    return true;
+                    // Give the user the chance to sort or filter the used servers for the query.
+                    DnsServerAddressStream serverList = parent.uncachedRedirectDnsServerStream(
+                            question.name(), nameServers);
+                    if (serverList != null) {
+                        query(serverList, 0, question,
+                              queryLifecycleObserver.queryRedirected(new DnsAddressStreamList(serverList)),
```

Comment:
future discussion: here is another discrepancy between list/stream like APIs that we should consider making consistent later.